### PR TITLE
Migrate to clipanion v3

### PR DESCRIPTION
**Pull request by Drarig29**
### What and why?

This refactor PR migrates clipanion from v2 to v3.

This will allow us to:
- Use `typanion` to validate **and coerce** the values of options, e.g. turn a string like `'0'`, `'false'`... into a boolean
- Use the `env` option to automatically resolve an option based on an environment variable
- Set a description for CLI arguments (before that, we could only document the usage of a command)
  - #865 

### How?

The main changes are:
- All `ExampleCommand.addOption()` are replaced by a private field set to `Option.*`
- All `ExampleCommand.addPath()` are replaced by `static paths = [ ... ]`
- In unit tests, constructing a command with `new ExampleCommand()` doesn't contain the default value or `undefined` anymore. You now need to use `createCommand(ExampleCommand)`, which will fix this for you.

I went with `Option.*` and not `@Option.*` annotations because the latter doesn't allow us to validate and coerce with `typanion`, and field types are not inferred.

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)

### CODE REVIEW

File: src/commands/lambda/__tests__/fixtures.ts [hunk 1]
```diff
@@ -47,28 +47,6 @@ export const makeCli = () => {
...
-
-/**
- * Allows to create an instance of any command that
- * extends the Command clss.
- *
- * @param commandClass any class that extends the Command class.
- * @param parameters parameters to use while creating the commandClass
- * @returns the instance of the given command with a mock context attatched.
- */
-export const createCommand = <T extends Command>(commandClass: ConstructorOf<T>, ...parameters: any[]) => {
```
[the remark this answers is not in the document]
Comment (duncanista) [reply src/commands/lambda/__tests__/fixtures.ts:64]: Nice, thanks!

File: src/commands/stepfunctions/uninstrument.ts [hunk 2]
```diff
@@ -26,8 +28,19 @@ export class UninstrumentStepFunctionsCommand extends Command {
...
+  // The options below are to match what InstrumentStepFunctionsCommand has so that customers can switch from instrument to uninstrument.
+  // Lambda command adopts the same approach as well.
+  private environment = Option.String('-e,--env', {hidden: true})
+  private forwarderArn = Option.String('--forwarder', {hidden: true})
+  private service = Option.String('--service', {hidden: true})
+  private mergeStepFunctionAndLambdaTraces = Option.Boolean(
+    '-mlt,--merge-lambda-traces,--merge-step-function-and-lambda-traces',
+    false,
+    {hidden: true}
+  )
```
Comment (duncanista) [inline src/commands/stepfunctions/uninstrument.ts:43]: Thanks for this piece of code, really appreciated!

**Review by duncanista — APPROVED** [review]
Looks good on the the Serverless side!

**Review by dastrong — APPROVED** [review]

File: src/commands/junit/upload.ts [hunk 3]
```diff
@@ -102,41 +105,38 @@ export class UploadJUnitXMLCommand extends Command {
...
+  })
+  private maxConcurrency = Option.String('--max-concurrency', '20', {validator: validation.isInteger()})
+  private metrics = Option.Array('--metrics')
+  private service = Option.String('--service', {env: 'DD_SERVICE'})
+  private tags = Option.Array('--tags')
+  private reportTags = Option.Array('--report-tags')
+  private reportMetrics = Option.Array('--report-metrics')
+  private rawXPathTags = Option.Array('--xpath-tag')
+  private gitRepositoryURL = Option.String('--git-repository-url')
+  private skipGitMetadataUpload = Option.String('--skip-git-metadata-upload', 'true', {validator: t.isBoolean()})
```
[the remark this answers is not in the document]
Comment (Drarig29) [reply src/commands/junit/upload.ts:125]: @juan-fernandez 

- for `--skip-git-metadata-upload`, proof that `0` and `false` are coerced into booleans: https://github.com/arcanis/typanion/blob/ab9f045afc697c998aff66992ad48cb8ba3478a4/sources/predicates/typePredicates.ts#L122-L132
- for `--logs`, here is some docs on `tolerateBoolean`: https://mael.dev/clipanion/docs/options#strings

According to my testing, preserved the different orders of precedence.

If you (or your team) prefer to handle this yourself, or not makes those changes at all, I can revert those changes from this PR if you want.

**Review by Letiste — APPROVED** [review]

**Review by xgouchet — APPROVED** [review]

File: src/commands/junit/upload.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Drarig29) [reply src/commands/junit/upload.ts:125]: When I talk about `env`, I talk about the argument to `Option.*`, not your `--env` flag.

I changed this:
```diff
-    if (!this.service) {
-      this.service = process.env.DD_SERVICE
-    }
+ private service = Option.String('--service', {env: 'DD_SERVICE'})
```

and this:
```diff
-    if (
-      !this.logs &&
-      process.env.DD_CIVISIBILITY_LOGS_ENABLED &&
-      !['false', '0'].includes(process.env.DD_CIVISIBILITY_LOGS_ENABLED.toLowerCase())
-    ) {
-      this.logs = true
-    }
+  private logs = Option.String('--logs', 'false', {
+    env: 'DD_CIVISIBILITY_LOGS_ENABLED',
+    tolerateBoolean: true,
+    validator: t.isBoolean(),
+  })
```

**Review by juan-fernandez — APPROVED** [review]
thanks for this! 🎉

**Comment by juan-fernandez** [comment]
I've added @andrea-mosk as reviewer of the `gate` command

File: src/commands/junit/upload.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Drarig29) [reply src/commands/junit/upload.ts:125]: cc. @juan-fernandez

**Review by andrea-mosk — APPROVED** [review]
Looks good from the `gate` command side

File: src/commands/synthetics/run-tests-command.ts [hunk 4]
```diff
@@ -44,25 +45,29 @@ export const DEFAULT_COMMAND_CONFIG: RunTestsCommandConfig = {
...
+
+  private apiKey = Option.String('--apiKey')
+  private appKey = Option.String('--appKey')
+  private datadogSite = Option.String('--datadogSite')
+  private failOnCriticalErrors = Option.Boolean('--failOnCriticalErrors')
+  private failOnMissingTests = Option.Boolean('--failOnMissingTests')
+  private failOnTimeout = Option.Boolean('--failOnTimeout')
+  private files = Option.Array('-f,--files')
+  private mobileApplicationVersionFilePath = Option.String('--mobileApp,--mobileApplicationVersionFilePath')
+  private pollingTimeout = Option.String('--pollingTimeout', {validator: validation.isInteger()})
```
Comment (jakepruitt) [inline src/commands/synthetics/run-tests-command.ts:62]: ❤️ love that validation

**Review by jakepruitt — APPROVED** [review]
Looks great!

**Pull request merged**